### ISSUE TITLE
bug(scheduler): consuming tasks in separate processes

### ISSUE THREAD

**Issue by jvancoillie** [Guikingone/SchedulerBundle#92]
Task is running multiple times.

with the following tasks configuration : 
```yaml
scheduler_bundle:
  transport:
    dsn: 'doctrine://default?execution_mode=first_in_first_out'
  tasks:
    foo:
      type: 'command'
      command: 'app:foo'
      description: '3mins task'
      single_run: true
    bar:
      type: 'command'
      command: 'app:bar'
      description: '1mins task'
      single_run: true
```

and a cron configuration : `* * * * * cd /path-to-your-project && php bin/console scheduler:consume >> /dev/null 2>&1`

the first `scheduler:consume` output :
```bash
   [OK] Task "bar" succeed. (Duration: 1 min, Memory used: 28.0 MiB)
```
one minute later second call `scheduler:consume` output :
```bash
   [OK] Task "foo" succeed. (Duration: 3 mins, Memory used: 27.9 MiB)
   [OK] Task "bar" succeed. (Duration: 1 min, Memory used: 28.0 MiB)
```

we can see that the `bar` task is executed twice. How can we ensure the uniqueness of tasks.

I hope to be clear enough in my explanation 😄

**Comment by Guikingone** [comment]
Regarding this issue, both `foo` and `bar` tasks have an `* * * * *` expression? 

So IMHO, it's normal that each task is executed when the command is launched BUT, it seems that the `single_run` directive is not used 🤔

**Comment by jvancoillie** [comment]
I agree with the single_run. but I think that on tasks performed monthly we could have the same behavior. 

```yaml
scheduler_bundle:
  transport:
    dsn: 'doctrine://default?execution_mode=first_in_first_out'
  tasks:
    foo:
      type: 'command'
      command: 'app:foo'
      description: '3mins task'
      expression: @monthly
    bar:
      type: 'command'
      command: 'app:bar'
      description: '1mins task'
      expression: @monthly
```
first `scheduler:consume` call :
```php
// dueTasks list
[ 0 => [
    'name' => 'foo',
    'state' => null,
    ],
  1 => [
    'name' => 'bar',
    'state' => null,
  ]    
]
``` 

second `scheduler:consume`  call : 
 ```php
 // dueTasks list
[ 0 => [
    'name' => foo, // locked by first consume call
    'state' => null, 
    ],
  1 => [
    'name' => 'bar',
    'state' => null,
  ]    
]
``` 
Here,  the `foo` task  will skip (lock by the locker in the first consume) it will handle the `bar` task.
When `bar` task is done the `worker` try again `foo`who now can be acquiere without retrieving done information of the first consume, so it run again ...

**Comment by Guikingone** [comment]
Hi @jvancoillie 👋🏻 

Could you test this issue against `0.4.4`? 

Thanks 🙂

**Comment by jvancoillie** [comment]
if I am not wrong, it is worse 😢 

```yaml
scheduler_bundle:
  transport:
    dsn: 'doctrine://default?execution_mode=first_in_first_out'
  tasks:
    foo:
      type: 'command'
      command: 'app:foo'
      description: '3mins task'
      single_run: true
    bar:
      type: 'command'
      command: 'app:bar'
      description: '1mins task'
      single_run: true
```

the first `scheduler:consume` output :
```bash
  [OK] Task "foo" succeed. (Duration: 3 mins, Memory used: 27.8 MiB)
  [OK] Task "bar" succeed. (Duration: 1 min, Memory used: 28.2 MiB)
```

second call `scheduler:consume` output :
```bash
  [OK] Task "bar" succeed. (Duration: 1 min, Memory used: 27.8 MiB)
  [OK] Task "bar" succeed. (Duration: 2 mins, Memory used: 28.2 MiB)
``` 

the `bar` task is played three times !

**Comment by Guikingone** [comment]
Hum, strange, can you give me more informations about the `bar` task? 

It is a task that take 3 minutes to be executed?

**Comment by jvancoillie** [comment]
yes these are two commands to test the behavior of the scheduler.

BarCommand.php
```php
   protected function execute(InputInterface $input, OutputInterface $output): int
    {
        $io = new SymfonyStyle($input, $output);
        $sleep = 1;
        $length = 60;
        $duration = $sleep * $length / 60;

        $io->success(sprintf('simple bar placeholder command (%d %s)', $duration, ($duration>1)?'mins':'min' ));

        $progressBar = new ProgressBar($output, $length);
        $progressBar->start();

        $i = 0;
        while ($i++ < $length) {
            $progressBar->advance();
            sleep($sleep);
        }

        $progressBar->finish();
        $io->writeln('');

        return Command::SUCCESS;
    }
```    
FooCommand.php
```php
    protected function execute(InputInterface $input, OutputInterface $output): int
    {
        $io = new SymfonyStyle($input, $output);
        $sleep = 3;
        $length = 60;
        $duration = $sleep * $length / 60;

        $io->success(sprintf('simple foo placeholder command (%d %s)', $duration, ($duration>1)?'mins':'min' ));

        $progressBar = new ProgressBar($output, $length);
        $progressBar->start();

        $i = 0;
        while ($i++ < $length) {
            $progressBar->advance();
            sleep($sleep);
            $this->em->clear();
        }

        $progressBar->finish();
        $io->writeln('');

        return Command::SUCCESS;
    }
```

**Comment by Guikingone** [comment]
Small question about this issue, the `scheduler:consume` is launched every minute? Via a CRON? Other?

**Comment by jvancoillie** [comment]
it will, but for testing I launch it manually

**Comment by Guikingone** [comment]
Hum, ok, can you try it using `scheduler:consume --wait`?

It seems that when running `schedumer:consume`, we don't check the current date, sadly, tasks are executed even if we're not on the next minute.

**Comment by jvancoillie** [comment]
one less `bar` task and one more `foo`task with `--wait` option 🤯 
<img width="1330" alt="Capture d’écran 2021-05-06 à 17 21 45" src="https://user-images.githubusercontent.com/21243943/117323950-a400bc00-ae8f-11eb-9392-986a52504ff3.png">

**Comment by Guikingone** [comment]
Quite strange that the first stopped 🤔 

> the first command has stopped and the second continues to perform tasks without stopping

The first should not stop until you use `CTRL + C`, have you used it?

Regarding the second one, I see that the tasks are executed then there's a "blank space" (so the `bar` execution I guess?) then it execute again both tasks, it looks like `bar` is not paused (that's a `single_run` one if I'm right?) 🤔 

> there is also an issue on the runtime duration

I'm not sure to see it 🤔

**Comment by jvancoillie** [comment]
> Quite strange that the first stopped 🤔
> 
Agree
> 
> The first should not stop until you use `CTRL + C`, have you used it?
> 
I don't think I did that, I have a doubt now. i will retest

> Regarding the second one, I see that the tasks are executed then there's a "blank space" (so the `bar` execution I guess?) then it execute again both tasks, it looks like `bar` is not paused (that's a `single_run` one if I'm right?) 🤔
> 
the blank spaces come from me I made a line break it was for the screen shot. It was then that I realized that the tasks continued
> 
> I'm not sure to see it 🤔

I have voluntarily left it running.

<img width="644" alt="Capture d’écran 2021-05-06 à 19 13 06" src="https://user-images.githubusercontent.com/21243943/117338899-6572fd80-ae9f-11eb-982d-2c3a6d32ea03.png">
we can see that the duration of the task `bar`is one hour it's supposed to run 1 minute

**Comment by jvancoillie** [comment]
I just retested and I confirm that I have not `CTRL-C`

**Comment by Guikingone** [comment]
> we can see that the duration of the task `foo` is one hour it's supposed to run 3 minutes

I think this is a display issue, I'm gonna investigate.

**Comment by Guikingone** [comment]
Hi @jvancoillie 👋🏻 

Could you test your issues against `dev-issue/92` when you have a moment? Thanks 🙂

**Comment by jvancoillie** [comment]
:( same result

**Comment by Guikingone** [comment]
Hi @jvancoillie 👋🏻 

I've tried to reproduce the situation you're facing and I'm close to the conclusion that the issue comes from the tasks you've defined and not from the worker (at least, not from the execution process), for exemple, here's the process I've followed: 

- Defining a set of tasks: 

```yaml
scheduler_bundle:
  transport:
    dsn: 'doctrine://default?execution_mode=first_in_first_out'
  tasks:
    foo:
      type: 'command'
      command: 'cache:clear'
      expression: '* * * * *'
      description: 'A simple cache clear task'
      options: ['--no-warmup']
    bar:
      type: 'chained'
      tasks:
        chained_foo:
          type: 'command'
          command: 'debug:container'
          expression: '* * * * *'
          description: 'A simple cache clear task'
          options: ['--no-warmup']
        chained_random:
          type: 'command'
          command: 'debug:container'
          expression: '* * * * *'
          description: 'A simple cache clear task'
          options: ['--no-warmup', '-vvv']
    list_directory:
      type: 'shell'
      command: ['ls', '-al']
      description: 'A simple command that list the current directories'
    list_directory_single:
      type: 'shell'
      command: ['ls', '-al']
      description: 'A simple command that list the current directories'
      single_run: true
```

- Running the command `scheduler:consume --wait`: 

<img width="939" alt="Screenshot 2021-05-08 at 08 02 55" src="https://user-images.githubusercontent.com/13744329/117528769-3ad69100-afd4-11eb-8b52-9684375646a7.png">

_Do not care about `44`, simple test for command execution_ 😄 

Every minutes, the tasks are retrieved and executed (in a blocking way if `bar` doesn't end at time for example). 

Here's when it becomes interesting, when using your two tasks, the `scheduler:consume --wait` fails (sometimes it stop, sometimes it block the whole worker), I'm not sure about the reason behind but can you investigate the `scheduler:consume` usage with "Symfony" defined commands and see if it works without your two "custom" tasks? 

Thanks for the feedback 🙂

**Comment by jvancoillie** [comment]
Hello @Guikingone, 

Hope your are doing well,

I just tested with symfony commands as requested on `issue/92`

Defined like this (i've kept the `single_run`)

```yaml
scheduler_bundle:
  transport:
    dsn: 'doctrine://default?execution_mode=first_in_first_out'
  tasks:
    foo:
      type: 'command'
      command: 'debug:autowiring'
      description: 'List classes/interfaces you can use for autowiring'
      single_run: true
    bar:
      type: 'command'
      command: 'debug:router'
      description: 'Display current routes for an application'
      single_run: true
```
 <img width="1422" alt="Capture d’écran 2021-05-08 à 08 40 22" src="https://user-images.githubusercontent.com/21243943/117529757-b25aef00-afd9-11eb-857f-35935866617b.png"> 
 
with this configuration, the execution on the other side is not testable because the execution time of the first worker is less than 1 minute.

However, the first execution continues to execute the tasks which are supposed to be launched only once :(

**Comment by Guikingone** [comment]
Ok, so, there's 2 issues: 

- The `single_run` is not used and the task runs even if executed before.

But, does the scheduler find tasks every minutes of launch the process without waiting?

**Comment by jvancoillie** [comment]
here are the logs I have when the tasks are repeated

```bash
[2021-05-08T11:06:19.672333+00:00] doctrine.DEBUG: SELECT COUNT(DISTINCT t.id) FROM _symfony_scheduler_tasks t WHERE t.task_name = :name LOCK IN SHARE MODE {":name":"foo"} []
[2021-05-08T11:06:19.711556+00:00] app.INFO: A task has been scheduled {"task":"foo"} []
[2021-05-08T11:06:19.712800+00:00] doctrine.DEBUG: SELECT COUNT(DISTINCT t.id) FROM _symfony_scheduler_tasks t WHERE t.task_name = :name LOCK IN SHARE MODE {":name":"bar"} []
[2021-05-08T11:06:19.714724+00:00] app.INFO: A task has been scheduled {"task":"bar"} []
[2021-05-08T11:06:19.770029+00:00] doctrine.DEBUG: SELECT COUNT(DISTINCT t.id) FROM _symfony_scheduler_tasks t LOCK IN SHARE MODE [] []
[2021-05-08T11:06:19.772065+00:00] doctrine.DEBUG: "START TRANSACTION" [] []
[2021-05-08T11:06:19.773222+00:00] doctrine.DEBUG: SELECT t.* FROM _symfony_scheduler_tasks t [] []
[2021-05-08T11:06:19.942562+00:00] doctrine.DEBUG: "COMMIT" [] []
[2021-05-08T11:06:19.949187+00:00] doctrine.DEBUG: SELECT COUNT(DISTINCT t.id) FROM _symfony_scheduler_tasks t LOCK IN SHARE MODE [] []
[2021-05-08T11:06:19.950819+00:00] doctrine.DEBUG: "START TRANSACTION" [] []
[2021-05-08T11:06:19.952238+00:00] doctrine.DEBUG: SELECT t.* FROM _symfony_scheduler_tasks t [] []
[2021-05-08T11:06:19.955832+00:00] doctrine.DEBUG: "COMMIT" [] []
[2021-05-08T11:06:19.957399+00:00] doctrine.DEBUG: "START TRANSACTION" [] []
[2021-05-08T11:06:19.958500+00:00] doctrine.DEBUG: DELETE FROM _symfony_scheduler_tasks [] []
[2021-05-08T11:06:19.959821+00:00] doctrine.DEBUG: "COMMIT" [] []
[2021-05-08T11:06:29.920886+00:00] doctrine.DEBUG: SELECT COUNT(DISTINCT t.id) FROM _symfony_scheduler_tasks t WHERE t.task_name = :name LOCK IN SHARE MODE {":name":"foo"} []
[2021-05-08T11:06:29.942022+00:00] doctrine.DEBUG: "START TRANSACTION" [] []
[2021-05-08T11:06:29.960711+00:00] doctrine.DEBUG: INSERT INTO _symfony_scheduler_tasks (task_name, body) VALUES(:name, :body) {":name":"foo",":body":"{\"body\":{\"command\":\"debug: [...]"} []
[2021-05-08T11:06:29.962097+00:00] doctrine.DEBUG: "COMMIT" [] []
[2021-05-08T11:06:29.988019+00:00] app.INFO: A task has been scheduled {"task":"foo"} []
[2021-05-08T11:06:29.989733+00:00] doctrine.DEBUG: SELECT COUNT(DISTINCT t.id) FROM _symfony_scheduler_tasks t WHERE t.task_name = :name LOCK IN SHARE MODE {":name":"bar"} []
[2021-05-08T11:06:29.991383+00:00] doctrine.DEBUG: "START TRANSACTION" [] []
[2021-05-08T11:06:29.993117+00:00] doctrine.DEBUG: INSERT INTO _symfony_scheduler_tasks (task_name, body) VALUES(:name, :body) {":name":"bar",":body":"{\"body\":{\"command\":\"debug: [...]"} []
[2021-05-08T11:06:29.994625+00:00] doctrine.DEBUG: "COMMIT" [] []
[2021-05-08T11:06:29.998814+00:00] app.INFO: A task has been scheduled {"task":"bar"} []
[2021-05-08T11:06:30.055490+00:00] doctrine.DEBUG: SELECT COUNT(DISTINCT t.id) FROM _symfony_scheduler_tasks t LOCK IN SHARE MODE [] []
[2021-05-08T11:06:30.057455+00:00] doctrine.DEBUG: "START TRANSACTION" [] []
[2021-05-08T11:06:30.058916+00:00] doctrine.DEBUG: SELECT t.* FROM _symfony_scheduler_tasks t [] []
[2021-05-08T11:06:30.238045+00:00] doctrine.DEBUG: "COMMIT" [] []
[2021-05-08T11:06:30.264999+00:00] app.INFO: The worker has been started {"failedTasks":0,"lastExecutedTask":null} []
[2021-05-08T11:06:30.266498+00:00] doctrine.DEBUG: SELECT COUNT(DISTINCT t.id) FROM _symfony_scheduler_tasks t LOCK IN SHARE MODE [] []
[2021-05-08T11:06:30.268370+00:00] doctrine.DEBUG: "START TRANSACTION" [] []
[2021-05-08T11:06:30.269697+00:00] doctrine.DEBUG: SELECT t.* FROM _symfony_scheduler_tasks t [] []
[2021-05-08T11:06:30.272845+00:00] doctrine.DEBUG: "COMMIT" [] []
[2021-05-08T11:06:30.285089+00:00] app.INFO: The worker is currently running {"failedTasks":0,"lastExecutedTask":null,"idle":false} []
[2021-05-08T11:06:30.295822+00:00] app.INFO: The worker is currently running {"failedTasks":0,"lastExecutedTask":null,"idle":false} []
[2021-05-08T11:06:30.716552+00:00] app.INFO: A task has been executed {"task":"foo"} []
[2021-05-08T11:06:30.718707+00:00] doctrine.DEBUG: "START TRANSACTION" [] []
[2021-05-08T11:06:30.722794+00:00] doctrine.DEBUG: UPDATE _symfony_scheduler_tasks SET body = :body WHERE task_name = :name {":name":"foo",":body":"{\"body\":{\"command\":\"debug: [...]"} []
[2021-05-08T11:06:30.724908+00:00] doctrine.DEBUG: "COMMIT" [] []
[2021-05-08T11:06:30.727592+00:00] doctrine.DEBUG: SELECT COUNT(DISTINCT t.id) FROM _symfony_scheduler_tasks t WHERE t.task_name = :name LOCK IN SHARE MODE {":name":"foo"} []
[2021-05-08T11:06:30.730236+00:00] doctrine.DEBUG: "START TRANSACTION" [] []
[2021-05-08T11:06:30.731968+00:00] doctrine.DEBUG: SELECT t.* FROM _symfony_scheduler_tasks t WHERE t.task_name = :name LOCK IN SHARE MODE {":name":"foo"} []
[2021-05-08T11:06:30.735726+00:00] doctrine.DEBUG: "COMMIT" [] []
[2021-05-08T11:06:30.737292+00:00] doctrine.DEBUG: "START TRANSACTION" [] []
[2021-05-08T11:06:30.739347+00:00] doctrine.DEBUG: UPDATE _symfony_scheduler_tasks SET body = :body WHERE task_name = :name {":name":"foo",":body":"{\"body\":{\"command\":\"debug: [...]"} []
[2021-05-08T11:06:30.741007+00:00] doctrine.DEBUG: "COMMIT" [] []
[2021-05-08T11:06:30.746078+00:00] app.INFO: The worker is currently running {"failedTasks":0,"lastExecutedTask":"foo","idle":true} []
[2021-05-08T11:06:30.762841+00:00] app.INFO: The worker is currently running {"failedTasks":0,"lastExecutedTask":"foo","idle":false} []
[2021-05-08T11:06:30.763981+00:00] app.INFO: The worker is currently running {"failedTasks":0,"lastExecutedTask":"foo","idle":false} []
[2021-05-08T11:06:31.290613+00:00] app.INFO: A task has been executed {"task":"bar"} []
[2021-05-08T11:06:31.293772+00:00] doctrine.DEBUG: "START TRANSACTION" [] []
[2021-05-08T11:06:31.298109+00:00] doctrine.DEBUG: UPDATE _symfony_scheduler_tasks SET body = :body WHERE task_name = :name {":name":"bar",":body":"{\"body\":{\"command\":\"debug: [...]"} []
[2021-05-08T11:06:31.304347+00:00] doctrine.DEBUG: "COMMIT" [] []
[2021-05-08T11:06:31.315003+00:00] doctrine.DEBUG: SELECT COUNT(DISTINCT t.id) FROM _symfony_scheduler_tasks t WHERE t.task_name = :name LOCK IN SHARE MODE {":name":"bar"} []
[2021-05-08T11:06:31.318698+00:00] doctrine.DEBUG: "START TRANSACTION" [] []
[2021-05-08T11:06:31.321364+00:00] doctrine.DEBUG: SELECT t.* FROM _symfony_scheduler_tasks t WHERE t.task_name = :name LOCK IN SHARE MODE {":name":"bar"} []
[2021-05-08T11:06:31.328204+00:00] doctrine.DEBUG: "COMMIT" [] []
[2021-05-08T11:06:31.334042+00:00] doctrine.DEBUG: "START TRANSACTION" [] []
[2021-05-08T11:06:31.336892+00:00] doctrine.DEBUG: UPDATE _symfony_scheduler_tasks SET body = :body WHERE task_name = :name {":name":"bar",":body":"{\"body\":{\"command\":\"debug: [...]"} []
[2021-05-08T11:06:31.339004+00:00] doctrine.DEBUG: "COMMIT" [] []
[2021-05-08T11:06:31.351147+00:00] app.INFO: The worker is currently running {"failedTasks":0,"lastExecutedTask":"bar","idle":true} []
[2021-05-08T11:07:00.336878+00:00] app.INFO: The worker has been started {"failedTasks":0,"lastExecutedTask":null} []
[2021-05-08T11:07:00.338769+00:00] doctrine.DEBUG: SELECT COUNT(DISTINCT t.id) FROM _symfony_scheduler_tasks t LOCK IN SHARE MODE [] []
[2021-05-08T11:07:00.341429+00:00] doctrine.DEBUG: "START TRANSACTION" [] []
[2021-05-08T11:07:00.343579+00:00] doctrine.DEBUG: SELECT t.* FROM _symfony_scheduler_tasks t [] []
[2021-05-08T11:07:00.348812+00:00] doctrine.DEBUG: "COMMIT" [] []
[2021-05-08T11:07:00.350637+00:00] app.INFO: The following task "foo" is paused|disabled, consider enable it if it should be executed! {"name":"foo","expression":"* * * * *","state":"paused"} []
[2021-05-08T11:07:00.351804+00:00] app.INFO: The following task "bar" is paused|disabled, consider enable it if it should be executed! {"name":"bar","expression":"* * * * *","state":"paused"} []
[2021-05-08T11:07:00.356153+00:00] app.INFO: The worker has been stopped {"failedTasks":0,"lastExecutedTask":null} []
[2021-05-08T11:07:00.357385+00:00] app.INFO: The worker is currently running {"failedTasks":0,"lastExecutedTask":null,"idle":false} []
[2021-05-08T11:07:00.358448+00:00] app.INFO: The worker is currently running {"failedTasks":0,"lastExecutedTask":null,"idle":false} []
[2021-05-08T11:07:00.371560+00:00] app.INFO: A task has been executed {"task":"foo"} []
[2021-05-08T11:07:00.374551+00:00] doctrine.DEBUG: "START TRANSACTION" [] []
[2021-05-08T11:07:00.378435+00:00] doctrine.DEBUG: UPDATE _symfony_scheduler_tasks SET body = :body WHERE task_name = :name {":name":"foo",":body":"{\"body\":{\"command\":\"debug: [...]"} []
[2021-05-08T11:07:00.380880+00:00] doctrine.DEBUG: "COMMIT" [] []
[2021-05-08T11:07:00.389156+00:00] doctrine.DEBUG: SELECT COUNT(DISTINCT t.id) FROM _symfony_scheduler_tasks t WHERE t.task_name = :name LOCK IN SHARE MODE {":name":"foo"} []
[2021-05-08T11:07:00.390909+00:00] doctrine.DEBUG: "START TRANSACTION" [] []
[2021-05-08T11:07:00.392512+00:00] doctrine.DEBUG: SELECT t.* FROM _symfony_scheduler_tasks t WHERE t.task_name = :name LOCK IN SHARE MODE {":name":"foo"} []
[2021-05-08T11:07:00.395016+00:00] doctrine.DEBUG: "COMMIT" [] []
[2021-05-08T11:07:00.396180+00:00] doctrine.DEBUG: "START TRANSACTION" [] []
[2021-05-08T11:07:00.397933+00:00] doctrine.DEBUG: UPDATE _symfony_scheduler_tasks SET body = :body WHERE task_name = :name {":name":"foo",":body":"{\"body\":{\"command\":\"debug: [...]"} []
[2021-05-08T11:07:00.400939+00:00] doctrine.DEBUG: "COMMIT" [] []
[2021-05-08T11:07:00.404854+00:00] app.INFO: The worker is currently running {"failedTasks":0,"lastExecutedTask":"foo","idle":true} []
[2021-05-08T11:07:00.406078+00:00] app.INFO: The worker is currently running {"failedTasks":0,"lastExecutedTask":"foo","idle":false} []
[2021-05-08T11:07:00.407162+00:00] app.INFO: The worker is currently running {"failedTasks":0,"lastExecutedTask":"foo","idle":false} []
[2021-05-08T11:07:00.413571+00:00] app.INFO: A task has been executed {"task":"bar"} []
[2021-05-08T11:07:00.416406+00:00] doctrine.DEBUG: "START TRANSACTION" [] []
[2021-05-08T11:07:00.419914+00:00] doctrine.DEBUG: UPDATE _symfony_scheduler_tasks SET body = :body WHERE task_name = :name {":name":"bar",":body":"{\"body\":{\"command\":\"debug: [...]"} []
[2021-05-08T11:07:00.422812+00:00] doctrine.DEBUG: "COMMIT" [] []
[2021-05-08T11:07:00.425538+00:00] doctrine.DEBUG: SELECT COUNT(DISTINCT t.id) FROM _symfony_scheduler_tasks t WHERE t.task_name = :name LOCK IN SHARE MODE {":name":"bar"} []
[2021-05-08T11:07:00.427333+00:00] doctrine.DEBUG: "START TRANSACTION" [] []
[2021-05-08T11:07:00.429343+00:00] doctrine.DEBUG: SELECT t.* FROM _symfony_scheduler_tasks t WHERE t.task_name = :name LOCK IN SHARE MODE {":name":"bar"} []
[2021-05-08T11:07:00.433280+00:00] doctrine.DEBUG: "COMMIT" [] []
[2021-05-08T11:07:00.435159+00:00] doctrine.DEBUG: "START TRANSACTION" [] []
[2021-05-08T11:07:00.437622+00:00] doctrine.DEBUG: UPDATE _symfony_scheduler_tasks SET body = :body WHERE task_name = :name {":name":"bar",":body":"{\"body\":{\"command\":\"debug: [...]"} []
[2021-05-08T11:07:00.439780+00:00] doctrine.DEBUG: "COMMIT" [] []
[2021-05-08T11:07:00.443108+00:00] app.INFO: The worker is currently running {"failedTasks":0,"lastExecutedTask":"bar","idle":true} []
```

**Comment by jvancoillie** [comment]
I will study this dysfunction to try to bring more element because debugging for you in these conditions is not optimal

**Comment by jvancoillie** [comment]
for the repetition problem at the interrupt, it is due to the behavior of the sleep method https://www.php.net/manual/en/function.sleep.php

When `ctrl-c` while sleeping sleep return the number of second left to sleep then execute is run again.

by adding a control like this on the sleep we can avoid this behavior
```php
  if ($this->options['sleepUntilNextMinute']) {
      $state = sleep($this->getSleepDuration());
      if (0 !== $state) {
          break;
      }
      $this->execute($options);
  }
```

**Comment by jvancoillie** [comment]
@Guikingone 
I think I understood why the tasks never stop.

This comes from recursion, with each call to `execute ()` (with `--wait` enabled) the recursion is started but the parent exection remains active since we are in the while which is never interrupted . 

```php
   if ($this->options['sleepUntilNextMinute']) {
      sleep($this->getSleepDuration());
      
      $this->execute($options);
      break // force stopping current while here and let the recurse working
  }
``` 
at least the problem is identified, all that remains is to work on it.  

Question, what is the behavior you want on --wait when there are no more tasks to do?
should we keep the process active until a CTRL-C is requested (that's how I see it)

**Comment by Guikingone** [comment]
Hi @jvancoillie 👋🏻 

Nice work on investigating it, so: 

> for the repetition problem at the interrupt, it is due to the behavior of the sleep method https://www.php.net/manual/en/function.sleep.php

Nice catch on this one, I've pushed a fix for this one on `dev-issue/92`.

> This comes from recursion, with each call to `execute()` (with `--wait` enabled) the recursion is started but the parent execution remains active since we are in the while which is never interrupted.

That's a harder issue, I agree on the solution, when it comes to: 

> should we keep the process active until a CTRL-C is requested (that's how I see it)

That's a very good question, IMHO, trying to consume tasks when there's none is not a problem, we can found tasks during the next execution (using the `--wait` option) 🤔

**Comment by Guikingone** [comment]
@jvancoillie Could you test `dev-issue/92` when you have a moment? This way, we can close this PR is the problem is solved and move forward on the other ones, thanks 🙂

**Comment by jvancoillie** [comment]
I agree with that, there is no endless repetition of tasks on the `--wait`.

However, when all the tasks are consumed, the worker stops. (which is no longer the case in my proposition #104)

there remains the problem of concurrent processing the `single_run`task is executed by both `consume` processes but we can work on it in another PR

**Comment by Guikingone** [comment]
> However, when all the tasks are consumed, the worker stops. (which is no longer the case in my proposition #104)

Strange situation, locally, I don't have the issue, even if the tasks are consumed (and when using `--wait`), the worker wait 🤔 

> there remains the problem of concurrent processing the single_runtask is executed by both consume processes but we can work on it in another PR

If you have more informations on this one, don't hesitate to explain 🙂

**Comment by jvancoillie** [comment]
> Strange situation, locally, I don't have the issue, even if the tasks are consumed (and when using `--wait`), the worker wait 🤔 

this only happened as soon as there is a `single_run` task, otherwise it still waits, it happened here: 

https://github.com/Guikingone/SchedulerBundle/blob/d4708191754f16a66389a99946ad379ccfe3b4a6/src/Worker/Worker.php#L110-L112
when task is last and paused it go out of shouldStop while loop and stop the worker.

> If you have more informations on this one, don't hesitate to explain 🙂

we fall back into the behavior why this issue was created (https://github.com/Guikingone/SchedulerBundle/issues/92#issue-874817866). 
tasks that require more time than the every minute configuration of the cron `* * * * *`

**Comment by Guikingone** [comment]
> this only happened as soon as there is a single_run task, otherwise it still waits, it happened here:

```php
 if (end($tasks) === $task && !$this->checkTaskState($task)) { 
     break 2; 
 } 
```

I don't have this code locally: 

```php
if (array_key_last($tasks) === $index && !$this->checkTaskState($task)) {
    break 2;
}
```

_Be sure to rebase_ 😄

I've noticed that `end()` moves the cursor even if called in a condition, `array_key_last` don't. 

Regarding this bug, the behaviour seems valid if the condition is removed, I'm gonna work on it:

<img width="877" alt="Screenshot 2021-05-09 at 17 37 16" src="https://user-images.githubusercontent.com/13744329/117578146-34880800-b0ed-11eb-8a29-52b9a0e513f8.png">

Here's the output when moving from a `break 2;` to `continue;`

<img width="869" alt="Screenshot 2021-05-09 at 17 39 27" src="https://user-images.githubusercontent.com/13744329/117578200-83ce3880-b0ed-11eb-9d17-39b026266b5c.png">

**Comment by Guikingone** [comment]
Nice, I've pushed a new version on `dev-issue/92`, feel free to test it when you have a moment 🙂

**Comment by jvancoillie** [comment]
Perfect, the consume remains pending 👍 


As for the separate process, I don't think the lock is involved.

because the tasks are mounted in memory before it is processed

suddenly the second run does not have the information that the first run processed the task 

first `scheduler:consume` call AT 0 bring up this tasklist
```php
// dueTasks list
[ 0 => [
    'name' => 'foo', // 3 mins to execute
    'state' => null,
    ],
  1 => [
    'name' => 'bar',
    'state' => null,
  ]    
]
``` 

second `scheduler:consume`  call at +1min bring the same tasklist; 
 ```php
 // dueTasks list
[ 0 => [
    'name' => foo, // locked by first consume call so skipped
    'state' => null, 
    ],
  1 => [
    'name' => 'bar', // processing this one
    'state' => null,
  ]    
]
```
when second call is done ...  the first call kept in memory the `bar` Task which is performed by the second process but which cannot obtain this information. 

j'espère que c'est assez clair ;)

**Comment by Guikingone** [comment]
Yes, that's enough clear 😄 

That's strange as locally, I'm able to consume tasks on two processes without issues: 

<img width="1673" alt="Screenshot 2021-05-09 at 17 43 20" src="https://user-images.githubusercontent.com/13744329/117580810-0eb53000-b0fa-11eb-9f6d-ca5b67e8a39a.png">

The `bar` task is consumed by the second process, I'm gonna investigate this lock call 🙂

**Comment by jvancoillie** [comment]
I think the lock file is working fine.
what if you configure a single_run task?
it works because the execution time of a task is less than 1 minute in this way the list of tasks is refreshed at the next call.

But if you have a first task that takes more than a minute and a second `single_run` task.
The second process will execute the `single_run` (the first being discarded by the locker) this same task will always be pending in the first process which means that it will be executed by both processes

**Comment by Guikingone** [comment]
Hi @jvancoillie 👋🏻 

Here's the behaviour that I encounter when scheduling your "long execution" task listed before (called `app:bar` here):

<img width="1673" alt="Screenshot 2021-05-10 at 08 14 15" src="https://user-images.githubusercontent.com/13744329/117613805-b87ed600-b167-11eb-97d7-285a6a6bd850.png">

The task is correctly locked in the second process, executed then "paused" 🤔

**Comment by jvancoillie** [comment]
yes the execution time of a task is not high enough.

can you try with a `foo` task  that takes 3 minutes and a `bar` task that takes 1 minute ? you should behave the same as me.

**Comment by Guikingone** [comment]
Here's the behaviour that I'm facing, `bar_single` (1 minute execution) is executed then `foo_single` is executed in the first process (3 minutes long), `bar_single` is executed twice in the second process 🤔 

<img width="1674" alt="Screenshot 2021-05-10 at 08 48 26" src="https://user-images.githubusercontent.com/13744329/117617206-7efc9980-b16c-11eb-82b8-6ce9301143c2.png">

Very strange situation, I'm gonna investigate it via the logs and see if the middleware are correctly called 🤔

**Comment by jvancoillie** [comment]
I tried to schematize the thing
![worker](https://user-images.githubusercontent.com/21243943/117618031-a9028b80-b16d-11eb-9887-e642373d021d.png)

**Comment by jvancoillie** [comment]
I haven't looked at the database! I'll take a look at it at lunch break

**Comment by Guikingone** [comment]
Ok, I've probably found the reason for this weird situation, the worker is wrongly configured for lock usage, these might be the internal reason behind the overlapping issue.

**Comment by jvancoillie** [comment]
Perfect, if you found the reason 👍 

If I can do anything, don't hesitate.

**Comment by Guikingone** [comment]
Hum, no, have you installed `symfony/lock`?

**Comment by jvancoillie** [comment]
of course, 
```bash 
composer/package-versions-deprecated 1.11.99.1            Composer plugin that provides efficient ...
doctrine/annotations                 1.12.1               Docblock Annotations Parser
doctrine/cache                       1.11.0               PHP Doctrine Cache library is a popular ...
doctrine/collections                 1.6.7                PHP Doctrine Collections library that ad...
doctrine/common                      3.1.2                PHP Doctrine Common project is a library...
doctrine/dbal                        2.13.1               Powerful PHP database abstraction layer ...
doctrine/deprecations                v0.5.3               A small layer on top of trigger_error(E_...
doctrine/doctrine-bundle             2.3.2                Symfony DoctrineBundle
doctrine/doctrine-migrations-bundle  3.1.1                Symfony DoctrineMigrationsBundle
doctrine/event-manager               1.1.1                The Doctrine Event Manager is a simple P...
doctrine/inflector                   2.0.3                PHP Doctrine Inflector is a small librar...
doctrine/instantiator                1.4.0                A small, lightweight utility to instanti...
doctrine/lexer                       1.2.1                PHP Doctrine Lexer parser library that c...
doctrine/migrations                  3.1.2                PHP Doctrine Migrations project offer ad...
doctrine/orm                         2.8.4                Object-Relational-Mapper for PHP
doctrine/persistence                 2.1.0                The Doctrine Persistence project is a se...
doctrine/sql-formatter               1.1.1                a PHP SQL highlighting library
dragonmantank/cron-expression        v3.1.0               CRON for PHP: Calculate the next or prev...
egulias/email-validator              3.1.1                A library for validating emails against ...
friendsofphp/proxy-manager-lts       v1.0.3               Adding support for a wider range of PHP ...
guikingone/scheduler-bundle          dev-issue/92 06edde3 A Symfony bundle that allows to schedule...
laminas/laminas-code                 4.2.2                Extensions to the PHP Reflection API, st...
laminas/laminas-eventmanager         3.3.1                Trigger and listen to events within a PH...
laminas/laminas-zendframework-bridge 1.2.0                Alias legacy ZF class names to Laminas P...
monolog/monolog                      2.2.0                Sends your logs to files, sockets, inbox...
nikic/php-parser                     v4.10.5              A PHP parser written in PHP
phpdocumentor/reflection-common      2.2.0                Common reflection classes used by phpdoc...
phpdocumentor/reflection-docblock    5.2.2                With this component, a library can provi...
phpdocumentor/type-resolver          1.4.0                A PSR-5 based resolver of Class names, T...
psr/cache                            1.0.1                Common interface for caching libraries
psr/container                        1.1.1                Common Container Interface (PHP FIG PSR-11)
psr/event-dispatcher                 1.0.0                Standard interfaces for event handling.
psr/link                             1.0.0                Common interfaces for HTTP links
psr/log                              1.1.4                Common interface for logging libraries
sensio/framework-extra-bundle        v5.6.1               This bundle provides a way to configure ...
symfony/asset                        v5.2.7               Manages URL generation and versioning of...
symfony/browser-kit                  v5.2.7               Simulates the behavior of a web browser,...
symfony/cache                        v5.2.7               Provides an extended PSR-6, PSR-16 (and ...
symfony/cache-contracts              v2.4.0               Generic abstractions related to caching
symfony/config                       v5.2.7               Helps you find, load, combine, autofill ...
symfony/console                      v5.2.7               Eases the creation of beautiful and test...
symfony/css-selector                 v5.2.7               Converts CSS selectors to XPath expressions
symfony/debug-bundle                 v5.2.4               Provides a tight integration of the Symf...
symfony/dependency-injection         v5.2.7               Allows you to standardize and centralize...
symfony/deprecation-contracts        v2.4.0               A generic function and convention to tri...
symfony/doctrine-bridge              v5.2.7               Provides integration for Doctrine with v...
symfony/dom-crawler                  v5.2.4               Eases DOM navigation for HTML and XML do...
symfony/dotenv                       v5.2.4               Registers environment variables from a ....
symfony/error-handler                v5.2.7               Provides tools to manage errors and ease...
symfony/event-dispatcher             v5.2.4               Provides tools that allow your applicati...
symfony/event-dispatcher-contracts   v2.4.0               Generic abstractions related to dispatch...
symfony/expression-language          v5.2.7               Provides an engine that can compile and ...
symfony/filesystem                   v5.2.7               Provides basic utilities for the filesystem
symfony/finder                       v5.2.4               Finds files and directories via an intui...
symfony/flex                         v1.12.2              Composer plugin for Symfony
symfony/form                         v5.2.7               Allows to easily create, process and reu...
symfony/framework-bundle             v5.2.7               Provides a tight integration between Sym...
symfony/http-client                  v5.2.7               Provides powerful methods to fetch HTTP ...
symfony/http-client-contracts        v2.4.0               Generic abstractions related to HTTP cli...
symfony/http-foundation              v5.2.7               Defines an object-oriented layer for the...
symfony/http-kernel                  v5.2.7               Provides a structured process for conver...
symfony/intl                         v5.2.7               Provides a PHP replacement layer for the...
symfony/lock                         v5.2.6               Creates and manages locks, a mechanism t...
symfony/mailer                       v5.2.7               Helps sending emails
symfony/maker-bundle                 v1.31.0              Symfony Maker helps you create empty com...
symfony/mime                         v5.2.7               Allows manipulating MIME messages
symfony/monolog-bridge               v5.2.7               Provides integration for Monolog with va...
symfony/monolog-bundle               v3.7.0               Symfony MonologBundle
symfony/notifier                     v5.2.4               Sends notifications via one or more chan...
symfony/options-resolver             v5.2.4               Provides an improved replacement for the...
symfony/phpunit-bridge               v5.2.7               Provides utilities for PHPUnit, especial...
symfony/polyfill-intl-grapheme       v1.22.1              Symfony polyfill for intl's grapheme_* f...
symfony/polyfill-intl-icu            v1.22.1              Symfony polyfill for intl's ICU-related ...
symfony/polyfill-intl-idn            v1.22.1              Symfony polyfill for intl's idn_to_ascii...
symfony/polyfill-intl-normalizer     v1.22.1              Symfony polyfill for intl's Normalizer c...
symfony/polyfill-mbstring            v1.22.1              Symfony polyfill for the Mbstring extension
symfony/polyfill-php73               v1.22.1              Symfony polyfill backporting some PHP 7....
symfony/polyfill-php80               v1.22.1              Symfony polyfill backporting some PHP 8....
symfony/process                      v5.2.7               Executes commands in sub-processes
symfony/property-access              v5.2.4               Provides functions to read and write fro...
symfony/property-info                v5.2.7               Extracts information about PHP class' pr...
symfony/proxy-manager-bridge         v5.2.4               Provides integration for ProxyManager wi...
symfony/routing                      v5.2.7               Maps an HTTP request to a set of configu...
symfony/security-bundle              v5.2.7               Provides a tight integration of the Secu...
symfony/security-core                v5.2.7               Symfony Security Component - Core Library
symfony/security-csrf                v5.2.7               Symfony Security Component - CSRF Library
symfony/security-guard               v5.2.4               Symfony Security Component - Guard
symfony/security-http                v5.2.7               Symfony Security Component - HTTP Integr...
symfony/serializer                   v5.2.7               Handles serializing and deserializing da...
symfony/service-contracts            v2.4.0               Generic abstractions related to writing ...
symfony/stopwatch                    v5.2.7               Provides a way to profile code
symfony/string                       v5.2.6               Provides an object-oriented API to strin...
symfony/translation                  v5.2.7               Provides tools to internationalize your ...
symfony/translation-contracts        v2.4.0               Generic abstractions related to translation
symfony/twig-bridge                  v5.2.7               Provides integration for Twig with vario...
symfony/twig-bundle                  v5.2.4               Provides a tight integration of Twig int...
symfony/validator                    v5.2.7               Provides tools to validate values
symfony/var-dumper                   v5.2.7               Provides mechanisms for walking through ...
symfony/var-exporter                 v5.2.7               Allows exporting any serializable PHP da...
symfony/web-link                     v5.2.5               Manages links between resources
symfony/web-profiler-bundle          v5.2.7               Provides a development tool that gives d...
symfony/workflow                     v5.2.7               Provides tools for managing a workflow o...
symfony/yaml                         v5.2.7               Loads and dumps YAML files
twig/extra-bundle                    v3.3.0               A Symfony bundle for extra Twig extensions
twig/twig                            v3.3.0               Twig, the flexible, fast, and secure tem...
webmozart/assert                     1.10.0               Assertions to validate method input/outp...

**Comment by Guikingone** [comment]
Ok, sorry, error on my side, fixed 🙂

**Comment by jvancoillie** [comment]
`issue/92` same behavior 

`bar` is executed twice 

<img width="1452" alt="Capture d’écran 2021-05-09 à 19 18 48" src="https://user-images.githubusercontent.com/21243943/117665669-5d6ad480-b1a3-11eb-8444-6b79dfedf276.png">

**Comment by jvancoillie** [comment]
for me the behavior is normal and the locker is not involved at least not in the current configuration. 
If we do not want there to be a duplicate, all the task list should be locked.

In this case the `bar` task is available and not blocked by the locker at the time of its request for execution.
here https://github.com/Guikingone/SchedulerBundle/issues/92#issuecomment-836263336 we can see that `bar`is never in lock state when looking for runners.

**Comment by Guikingone** [comment]
Yes, I was thinking about locking the whole task list, I'm gonna investigate this solution, that's strange as I don't face this issue locally 🙁

**Comment by Guikingone** [comment]
> Yes it's really strange that you can't reproduce this problem locally.
>
> Are you still on a configuration with 3mins and 1min single_run tasks ?

Yes

> Maybe not necessary to lock the whole list, only lock the single_run? 

That's the issue, actually, each task is locked just before being executed, which seemed fine until now, that's very strange that only `single_run` trigger this behaviour 🤔

**Comment by jvancoillie** [comment]
I understand why this happens on the single_run;

Couldn't the lock as it exists today have consequences if we want to set up a priority on the sequence of tasks, you will tell me we can always go through the chainedType to make sure that!

**Comment by jvancoillie** [comment]
yes i agree with that.

but with the multiple run we can expect the following configuration:
```yaml
scheduler_bundle:
  transport:
    dsn: 'doctrine://default?execution_mode=first_in_first_out'
  tasks:
      bar:
          type: 'command'
          command: 'app:bar'
          expression: '@monthly'
          description: '3 minutes task'
          priority: 1
      foo:
          type: 'command'
          command: 'app:foo'
          expression: '@monthly'
          description: '1 minute task'
          priority: 2
 ```
result in 
the first `scheduler:consume` output :
```bash
   [OK] Task "bar" succeed. (Duration: 3 min, Memory used: 28.0 MiB)
```
one minute later second call `scheduler:consume` output :
```bash
   [OK] Task "foo" succeed. (Duration: 1 min, Memory used: 28.0 MiB)
``` 
Priority 2 task will be completed before priority 1 task

nothing bad, the execution order is respected 👍 


it should be understood that the priority does not mean that the task will be executed one after the other (no dependencies chainedTask for that 😉) but can be executed in parallel with a delayed start compared to the priority order

**Comment by Guikingone** [comment]
If tasks needs priority support during execution, that's an other topic (not related to the current lock one) to explore 🙂

**Comment by Guikingone** [comment]
For now, I prefer to fix the lock issue then merge it, this way, we can safely explore improvements built on this patch 🙂

**Comment by Guikingone** [comment]
> I won't bother you any longer with my wacky thoughts

No problem here, maybe the issue is not the best place to discuss about this type of "improvement" (consider opening a specific discussion) but I'm always happy to discuss about improvements 🙂

**Comment by jvancoillie** [comment]
yes no problem, for me it was linked to the fact of locking the entire task list but you're right we must move forward

**Comment by Guikingone** [comment]
Hi @jvancoillie 👋🏻 

Small update on this issue, I'm currently working on a fix via `dev-issue/92` if you want to test it, the limitation was related to the lock store used (by default, the `fs` one does not support serializing the key which is a required things to handle multiple process execution). 

It should be ready for `0.6` but if you want to test it before, feel free to 🙂

**Comment by jvancoillie** [comment]
Hello @Guikingone, 

with great pleasure, a little complicated lately but i will take a look at it as soon as possible

**Comment by jvancoillie** [comment]
hi @Guikingone 👋 

on composer update to `dev-issue/92`i got the following errror :
```bash
Executing script cache:clear [KO]
 [KO]
Script cache:clear returned with error code 1
!!
!!  In SchedulerBundleConfiguration.php line 34:
!!
!!    Warning: count(): Parameter must be an array or an object that implements C
!!    ountable
!!
!!
!!
Script @auto-scripts was called via post-update-cmd
```

**Comment by jvancoillie** [comment]
And on `bin/console scheduler:reboot`
```bash
bash-5.1# bin/console scheduler:reboot

In Connection.php line 219:

  The given task cannot be updated as the identifier or the body is invalid
```

are there any configuration changes to be expected. I will try to understand.

**Comment by Guikingone** [comment]
Hum strange, I'm gonna investigate these issues 🤔

**Comment by jvancoillie** [comment]
I just took a look. it seems that the key cannot be serialized, it is marked as non-serializable
```bash
^ SchedulerBundle\TaskBag\LockTaskBag^ {#549
  -key: Symfony\Component\Lock\Key^ {#544
    -resource: "_symfony_scheduler__bar_2021_06_20_01"
    -expiringTime: null
    -state: array:1 [
      "Symfony\Component\Lock\Store\FlockStore" => array:2 [
        0 => false
        1 => stream resource {@1749
          timed_out: false
          blocked: true
          eof: false
          wrapper_type: "plainfile"
          stream_type: "STDIO"
          mode: "r+"
          unread_bytes: 0
          seekable: true
          uri: "/tmp/sf._symfony_scheduler__bar_2021_06_20_01.6LZ4Egp.lock"
          options: []
        }
      ]
    ]
    -serializable: false
  }
}
```
https://github.com/Guikingone/SchedulerBundle/blob/a93b706084667a81b47439dd130fad8526d560e2/src/Serializer/LockTaskBagNormalizer.php#L37

which throws an exception `The key can not be serialized.`

**Comment by jvancoillie** [comment]
Indeed it is the `FlockStore` which is incompatible with serialization :(

https://symfony.com/doc/current/components/lock.html#serializing-locks

**Comment by jvancoillie** [comment]
Another point, the `rowCount` on a [MySQL update](https://www.php.net/manual/fr/pdostatement.rowcount.php#104930) without change returns 0 which triggers a false positive on the update.
https://github.com/Guikingone/SchedulerBundle/blob/6dc24eb86ab016d147dbfd1a806a2bd2c71cb143/src/Bridge/Doctrine/Transport/Connection.php#L214

**Comment by Guikingone** [comment]
Hi @jvancoillie 👋🏻 

Yes, the `FlockStore` cannot support the serialisation of `Key`, that's why the PR introduce a BC break. 

Regarding: 

> Another point, the rowCount on a MySQL update without change returns 0 which triggers a false positive on the update.

Could you open a new issue?

**Comment by jvancoillie** [comment]
no problem, i will open a new pr for the mysql update problem.

Concerning the `lock_store` semaphore works well, maybe it would be interesting to use the default configuration of the component lock ;) #155

**Comment by jvancoillie** [comment]
Otherwise the PR #102 seems correct, really nice work 👍 

apart from the little doctrine update problem #157 and I'm also wondering if the finally block wouldn't overwrite the `ExecutionLockBag` information https://github.com/Guikingone/SchedulerBundle/pull/102#discussion_r654941146
